src/ripple/app/tx/impl/NFTokenAcceptOffer.cpp

```diff
@@ -0,0 +1,366 @@
+            return tecINSUFFICIENT_FUNDS;
+    }
+
+    if (sell)
+    {
+        auto const so = ctx.view.read(keylet::nftoffer(*sell));
+
+        if (((*so)[sfFlags] & lsfSellToken) != lsfSellToken)
+            return tecOFFER_TYPE_MISMATCH;
+
+        // An account can't accept an offer it placed:
+        if ((*so)[sfOwner] == ctx.tx[sfAccount])
+            return tecCANT_ACCEPT_OWN_OFFER;
+
+        // The seller must own the token.
+        if (!nft::findToken(ctx.view, (*so)[sfOwner], (*so)[sfTokenID]))
+            return tecNO_PERMISSION;
+
+        // If not in bridged mode...
+        if (!buy)
+        {
+            // If the offer has a Destination field, the acceptor must be the
+            // Destination.
+            if (auto const dest = so->at(~sfDestination);
+                dest.has_value() && *dest != ctx.tx[sfAccount])
+                return tecNO_PERMISSION;
+        }
+
+        // The account offering to buy must have funds:
+        auto const needed = so->at(sfAmount);
+
+        if (accountHolds(
+                ctx.view,
+                ctx.tx[sfAccount],
+                needed.getCurrency(),
+                needed.getIssuer(),
+                fhZERO_IF_FROZEN,
+                ctx.j) < needed)
+            return tecINSUFFICIENT_FUNDS;
+    }
+
+    return tesSUCCESS;
+}
+
+TER
+NFTokenAcceptOffer::pay(
+    AccountID const& from,
+    AccountID const& to,
+    STAmount amount)
+{
+    // This should never happen, but it's easy and quick to check.
+    if (amount < beast::zero)
+        return tecINTERNAL;
+
+    return accountSend(view(), from, to, amount, j_);
+}
+
+TER
+NFTokenAcceptOffer::acceptOffer(std::shared_ptr<SLE> const& offer)
+{
+    bool const isSell = offer->isFlag(lsfSellToken);
+    AccountID const owner = (*offer)[sfOwner];
+    AccountID const& seller = isSell ? owner : account_;
+    AccountID const& buyer = isSell ? account_ : owner;
+
+    auto const tokenID = (*offer)[sfTokenID];
+
+    if (auto amount = offer->getFieldAmount(sfAmount); amount != beast::zero)
+    {
+        // Calculate the issuer's cut from this sale, if any:
+        if (auto const fee = nft::getTransferFee(tokenID); fee != 0)
+        {
+            auto const cut = multiply(amount, transferFeeAsRate(fee));
+
+            if (auto const issuer = nft::getIssuer(tokenID);
+                cut != beast::zero && seller != issuer && buyer != issuer)
+            {
+                if (auto const r = pay(buyer, issuer, cut); !isTesSuccess(r))
+                    return r;
+                amount -= cut;
+            }
+        }
+
+        // Send the remaining funds to the seller of the NFT
+        if (auto const r = pay(buyer, seller, amount); !isTesSuccess(r))
+            return r;
+    }
+
+    // Now transfer the NFT:
+    auto nft = nft::findToken(view(), seller, tokenID);
+
+    if (!nft)
+        return tecINTERNAL;
+
+    if (auto const ret = nft::removeToken(view(), seller, tokenID);
```

Comment:
As it stands right now, `nft::findToken()` can't do this work, because it takes a `ReadView` parameter instead of an `ApplyView`.  What I'll try is adding a new free function, `nft::findTokenAndPage`, that will take an `ApplyView` and return both the token object and a `std::shared_ptr<SLE>` to the page.  That may help.  And it may be useful elsewhere.  We'll see.